### ISSUE TITLE
Too many s3-transfer-manager-worker threads  #435

### ISSUE THREAD

**Issue by veerumallavarapu** [aws/aws-sdk-java#1463]
Hi All,

previous closed ticket : : https://github.com/aws/aws-sdk-java/issues/435

We are currently on 1.11.176, however we are experiencing the same issue as mentioned in the description. We are currently seeing 2949 instances of s3 transfer manager worker threads in the thread dump that we acquired. As a result of this, the CPU utilization is going high (almost 100%). Could you please check this again?

We are using the API as shown below.
public TransferManager createS3TransferManager(){
	TransferManagerBuilder transferManagerBuilder = TransferManagerBuilder.standard().withS3Client(s3Client);
	transferManagerBuilder.setShutDownThreadPools(false);
	return transferManagerBuilder.build();
}

private void upload(final TransferManager transferManager, final PutObjectRequest uploadObjRequest, final AmazonS3FileTransferImpl transfer){
	uploadObjRequest.setGeneralProgressListener(new ProgressListener() {
		@Override
		public void progressChanged(ProgressEvent progressEvent) {
			transfer.progress(progressEvent.getBytesTransferred());
		}
	});

	transfer.start();

	final Upload upload = transferManager.upload(uploadObjRequest);
	try{
		upload.waitForCompletion();
	}catch(AmazonServiceException e){
		throw new FSRuntimeException(e);
	}catch(InterruptedException ie){
		LOG.warn("Uploading content to Amazon S3 is interuppted: " + uploadObjRequest.getBucketName() + "-->" + uploadObjRequest.getKey());
		Thread.currentThread().interrupt();
	}finally{
		transfer.end();
		transferManager.shutdownNow(false);
	}
}

[threaddumps-P01.log](https://github.com/aws/aws-sdk-java/files/1700145/threaddumps-P01.log)

**Comment by varunnvs92** [comment]
If you are using default TransferManager, only 10 threads are created during construction.
https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/TransferManagerUtils.java#L54

Are you constructing multiple instances of TransferManager?

`transferManagerBuilder.setShutDownThreadPools(false);`
I doubt the issue might be at this line in your code. When transferManager is shutdown, we automatically close the executor threads associated with it. Setting this option to false disables that behavior. Can I know the reason for disabling this feature? You can try removing this and see if it improves the situation.

**Comment by viswanadhbysani** [comment]
Yes we are creating multiple instances of TransferManager depending on the number of the files that we are uploading. 

Also regarding the **transferManagerBuilder.setShutDownThreadPools(false)**, we were getting "Upload Canceled" error when we tried to upload large (>1 GB) hls files to S3. Hence we had to handle it by setting the shutdownThreadPools to false while creating the transferManager, and closing those threadPools using the transferManager.shutdownNow() in the finally method. I have attached the image of the same. 

![image](https://user-images.githubusercontent.com/24470701/35901348-2cd8df6a-0bfc-11e8-990f-2dda18a8226e.png)

The log file contains the Upload Canceled error that we were getting prior to controlling the threadPool shutdown.

[uploadCanceled.log](https://github.com/aws/aws-sdk-java/files/1701964/uploadCanceled.log)

Please let us know if you need any further information for debugging this case.

Thanks,
Kasi

**Comment by varunnvs92** [comment]
The error is thrown at this [line](https://github.com/aws/aws-sdk-java/blob/4734de6fb0f80fe5768a6587aad3b9d0eaec388f/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/UploadCallable.java#L249). This means your application is shutting down the thread pool somewhere in the application before the TransferManager (TM) shutdown is closed.

**Comment by viswanadhbysani** [comment]
Yup, it is being shutdown there. However no where in our code, we were shutting it down. It was getting shutdown automatically. So we had to control the shutdown behavior in order to get rid of Upload Canceled error. 

Thanks,
Kasi

**Comment by varunnvs92** [comment]
Thats weird. Were you using the default threadpool created by TransferManager or Did you pass a custom threadpool to TM during construction?

**Comment by viswanadhbysani** [comment]
We were using default threadpool created by TransferManager while constructing it. We were using the below code earlier (when upload canceled error was coming),

	public TransferManager createS3TransferManager(){
		return new TransferManager(s3Client);
	}

	private void upload(final TransferManager transferManager, final PutObjectRequest uploadObjRequest, final AmazonS3FileTransferImpl transfer){
		uploadObjRequest.setProgressListener(new ProgressListener() {
			@Override
			public void progressChanged(ProgressEvent progressEvent) {
				transfer.progress(progressEvent.getBytesTransfered());
			}
		});

		transfer.start();

		final Upload upload = transferManager.upload(uploadObjRequest);
		try{
			upload.waitForCompletion();
		}catch(AmazonServiceException e){
			throw new FSRuntimeException(e);
		}catch(InterruptedException ie){
			LOG.warn("Uploading content to Amazon S3 is interuppted: " + uploadObjRequest.getBucketName() + "-->" + uploadObjRequest.getKey());
			Thread.currentThread().interrupt();
		}finally{
			transfer.end();
		}
	}

Thanks,
Kasi

**Comment by dagnir** [comment]
To prevent the thread leaks, you definitely want to be shutting down the thread pools after you're done using the `TransferManager`.

By adding `transferManagerBuilder.setShutDownThreadPools(false)`, when you call `transferManager.shutdownNow(false);`, it essentially does nothing because neither the `TransferManager` threadpools or `AmazonS3Client` will be shut down:

https://github.com/aws/aws-sdk-java/blob/6d5b15168ece5238b6ba7764f703b3a720e91c81/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/TransferManager.java#L1713-L1722

Note that this code snippet for `shutdowNow(boolean)` is different from the code screenshot posted in https://github.com/aws/aws-sdk-java/issues/1463#issuecomment-363670514. The screenshot looks like it's from an older version of the SDK.

**Comment by viswanadhbysani** [comment]
Thanks @dagnir and @varunnvs92 for your time on this issue.

We found a problem with our code while uploading the HLS file (or any zip file) to amazon s3. Actually we are creating TransferManager object for each of the file that is present in the zip while copying it (and its contents) to S3. As a result of this, suppose if HLS zip file contains 1000 ts segments, 1000 TransferManager objects were getting created. We have rectified this problem, now we are making use of single TransferManger object for the entire zip file. 

Thanks again @varunnvs92 and @dagnir for your help.

Thanks,
Kasi
